CIPs/0010.md

```diff
@@ -29,15 +29,36 @@ As applications get developed on Celo, it becomes evident that this mechanism of
 
 Expand the `Signers` struct in `Accounts.sol` with a generic mapping and generic version of `authorizeXSigner`, `removeXSigner`, `getXSigner` and `signerToAccount` that take as additional input `string namespace`.
 
-- multiple account keys? (effectively making lockeedgold somewhat of a custody solution)
-- authorization key
-- makes key management easier by allowing tools implement once interface
+- Multiple account keys? (effectively making locked gold somewhat of a custody solution)
+- Authorization key
+- Makes key management easier by allowing tools implement once interface
 
 ## Alternative Solutions
 
-Alternately, applications could always derive their own authorization scheme/smart contracts. However, it would not benefit of the tooling that could be possible to faciliate key management.
+Alternately, applications could always derive their own authorization scheme/smart contracts. However, it would not benefit from the tooling that could be possible for facilitating key management.
 
-Instead of extending the `Signers` struct, a completely different smart contract `AuthorizedSigners` could be created. Functionality would be practically duplicated, but separated from the critical PoS system while gaining the benefit from a standardized authorizaton scheme.
+Instead of extending the `Signers` struct, a completely different smart contract `AuthorizedSigners` could be created. Functionality would be practically duplicated, but separated from the critical PoS system while gaining the benefit from a standardized authorization scheme.
+
+Alternative Solution: Enable smart contracts to be used as signers (as an addition to existing implementation)
+
+In the current implementation, it is not possible to authorize smart contract addresses as signers, since they do not have private keys and are unable to generate the proofs required. To account for that limitation, a "smart contract signer authorization" method can be added to the `Accounts` contract with the following characteristics:
+
+Note: Each signer role will have its own variant of the solution outlined below (e.g. `authorizeAttestationSignerContract`, `authorizeVoteSignerContract`, etc. - similar to the existing implementation) - until the plan for the generalized signer authorization method is finalized.
+
+- Restricts smart contracts that are authorizable as signers to those created by `msg.sender` (for now, to limit potential damage from unforeseen malicious behavior)
```

Comment:
I assume here that `msg.sender` is the account? What do you do if the account is a `ReleaseGold` smart contract (or any other smart contract)?

Reply:
That's correct. A smart contract would receive the same treatment as a "normal" account: if it is a registered/Locked Gold Celo account then we would proceed with verifying whether it had created the signer smart contract.

Reply:
Are we sure that this aids security? IMO, this will require the account key to deploy a contract which seems very difficult to most users to do

Reply:
Thanks for the thought-provoking questions @nambrot .

> Are we sure that this aids security?

Great question. After thinking about it further, since it's a common practice for contracts to create contracts and assign their ownership to others, verifying creatorship alone is probably not enough. Back to the drawing board!

> IMO, this will require the account key to deploy a contract which seems very difficult to most users to do 

Since most of the contract-deploying tx can be prepopulated (besides the account nonce) by the developer, it would be virtually zero effort for users - all that would be needed is their signature.